src/object.d

```diff
@@ -3340,6 +3349,14 @@ private void _doPostblit(T)(T[] arr)
     }
 }
 
+private void _doCopyCtor(T, U)(T[] src, U[] dst)
+{
+    foreach (i, ref elem; src)
+    {
+        dst[i].__ctor(elem);
```

Comment:
What if it throws? Or rather, what if it throws, exception is caught, and later GC collects the array? Will it call any dtors? All of them? On a partially-initialized array?

Reply:
Yah, this needs care:

```d
size_t i = 0;
scope (failure)
{
    for (size_t j = 0; j < i; ++j)
        dst[j].__dtor();
    delete dst;
}
for (; i < src.length; ++i)
    dst[i].__ctor(elem);
```

Reply:
Well not `delete` :)

Reply:
`__xdtor`. If exists. And probably, before this call, not a `_d_newarrayU`, but a `_d_newarrayiT`. ;)

Reply:
Direct call to __ctor doesn't seem to write .init. Bad things may happen (c).

Reply:
This should probably call [_d_arrayctor](https://github.com/dlang/druntime/blob/master/src/core/internal/array/construction.d#L24). Otherwise, we'll need to replicate the logic there as well. It seems that `_d_arrayctor` already implements the logic necessary to handle exceptions.

Also, `_d_arrayctor` uses [postblitRecurse](https://github.com/dlang/druntime/blob/master/src/core/internal/postblit.d#L28) which knows about the new copyCtor.

There is, however, an issue with `_d_arrayctor`: it does a [memcpy](https://github.com/dlang/druntime/blob/master/src/core/internal/array/construction.d#L56) before calling `postblitRecurse`, which defies the purpose of the copyCtor, but this is an easy fix (I'll open a PR for it)